actors/builtin/miner/expiration_queue_test.go

```diff
@@ -9,33 +15,629 @@ import (
+		assert.True(t, set.FaultyPower.Equals(PowerForSectors(sectorSize, sectors[2:])))
+
+		// expect last set to only contain non faulty sector
+		requireNoExpirationGroupsBefore(t, 13, queue)
+		set, err = queue.PopUntil(13)
+		require.NoError(t, err)
+
+		assertBitfieldEmpty(t, set.OnTimeSectors)
+		assertBitfieldEmpty(t, set.EarlySectors)
+
+		// all pledge is dropped
+		assert.Equal(t, big.Zero(), set.OnTimePledge)
+
+		assert.True(t, set.ActivePower.Equals(NewPowerPairZero()))
+		assert.True(t, set.FaultyPower.Equals(NewPowerPairZero()))
+	})
+
+	t.Run("reschedule recover restores all sector stats", func(t *testing.T) {
+		// Create expiration 3 sets with 2 sectors apiece
+		queue := emptyExpirationQueueWithQuantizing(t, QuantSpec{unit: 4, offset: 1})
+		_, _, _, err := queue.AddActiveSectors(sectors, sectorSize)
+		require.NoError(t, err)
+
+		_, err = queue.Root()
+		require.NoError(t, err)
+
+		// Fault middle sectors to expire at epoch 6 to put sectors in a state
+		// described in "reschedules sectors as faults"
+		_, err = queue.RescheduleAsFaults(abi.ChainEpoch(6), sectors[1:5], sectorSize)
+		require.NoError(t, err)
+
+		_, err = queue.Root()
+		require.NoError(t, err)
+
+		// mark faulted sectors as recovered
+		recovered, err := queue.RescheduleRecovered(sectors[1:5], sectorSize)
+		require.NoError(t, err)
+		assert.True(t, recovered.Equals(PowerForSectors(sectorSize, sectors[1:5])))
+
+		// expect first set to contain first two sectors with active power
+		requireNoExpirationGroupsBefore(t, 5, queue)
+		set, err := queue.PopUntil(5)
+		require.NoError(t, err)
+
+		assertBitfieldEquals(t, set.OnTimeSectors, 1, 2)
+		assertBitfieldEmpty(t, set.EarlySectors)
+
+		// pledge from both sectors
+		assert.Equal(t, big.NewInt(2001), set.OnTimePledge)
+
+		assert.True(t, set.ActivePower.Equals(PowerForSectors(sectorSize, sectors[:2])))
+		assert.True(t, set.FaultyPower.Equals(NewPowerPairZero()))
+
+		// expect second set to have lost early sector 5 and have active power just from 3 and 4
+		requireNoExpirationGroupsBefore(t, 9, queue)
+		set, err = queue.PopUntil(9)
+		require.NoError(t, err)
+
+		assertBitfieldEquals(t, set.OnTimeSectors, 3, 4)
+		assertBitfieldEmpty(t, set.EarlySectors)
+
+		// pledge is kept from original 2 sectors
+		assert.Equal(t, big.NewInt(2005), set.OnTimePledge)
+
+		assert.True(t, set.ActivePower.Equals(PowerForSectors(sectorSize, sectors[2:4])))
+		assert.True(t, set.FaultyPower.Equals(NewPowerPairZero()))
+
+		// expect sector 5 to be returned to last setu
+		requireNoExpirationGroupsBefore(t, 13, queue)
+		set, err = queue.PopUntil(13)
+		require.NoError(t, err)
+
+		assertBitfieldEquals(t, set.OnTimeSectors, 5, 6)
+		assertBitfieldEmpty(t, set.EarlySectors)
+
+		// Pledge from sector 5 is restored
+		assert.Equal(t, big.NewInt(2009), set.OnTimePledge)
+
+		assert.True(t, set.ActivePower.Equals(PowerForSectors(sectorSize, sectors[4:])))
+		assert.True(t, set.FaultyPower.Equals(NewPowerPairZero()))
```

Comment:
My mental summary of this state is that it should be just like it was before the fault. Could we make that comparison more directly?